File: system-test/dotnet/Driver.Adapter/External/Erp/Client/BaseErpClient.cs
using Driver.Adapter.Shared.Client.Http;

using Common;

using Driver.Adapter.External.Erp.Client.Dtos;

using Driver.Adapter.External.Erp.Client.Dtos.Error;



namespace Driver.Adapter.External.Erp.Client;



public abstract class BaseErpClient : IDisposable

{

    private const string HealthEndpoint = "/health";

    private const string ProductsEndpoint = "/api/products";



    protected readonly JsonHttpClient<ExtErpErrorResponse> HttpClient;

    private bool _disposed;



    protected BaseErpClient(string baseUrl)

    {

        HttpClient = new JsonHttpClient<ExtErpErrorResponse>(baseUrl);

    }



    public void Dispose()

    {

        Dispose(true);

        GC.SuppressFinalize(this);

    }



    protected virtual void Dispose(bool disposing)

    {

        if (_disposed) return;

        if (disposing)

            HttpClient?.Dispose();

        _disposed = true;

    }



    public Task<Result<VoidValue, ExtErpErrorResponse>> CheckHealthAsync()

        => HttpClient.GetAsync(HealthEndpoint);



    public Task<Result<ExtProductDetailsResponse, ExtErpErrorResponse>> GetProductAsync(string? sku)

        => HttpClient.GetAsync<ExtProductDetailsResponse>($"{ProductsEndpoint}/{sku}");

}

File: system-test/java/src/main/java/com/mycompany/myshop/testkit/driver/adapter/external/erp/client/BaseErpClient.java
package com.mycompany.myshop.testkit.driver.adapter.external.erp.client;

import com.mycompany.myshop.testkit.driver.adapter.external.erp.client.dtos.ExtProductDetailsResponse;
import com.mycompany.myshop.testkit.driver.adapter.external.erp.client.dtos.error.ExtErpErrorResponse;
import com.mycompany.myshop.testkit.driver.adapter.shared.client.http.JsonHttpClient;
import com.mycompany.myshop.testkit.common.Closer;
import com.mycompany.myshop.testkit.common.Result;

public abstract class BaseErpClient implements AutoCloseable {
    private static final String HEALTH_ENDPOINT = "/health";
    private static final String PRODUCTS_ENDPOINT = "/api/products";

    protected final JsonHttpClient<ExtErpErrorResponse> httpClient;

    protected BaseErpClient(String baseUrl) {
        this.httpClient = new JsonHttpClient<>(baseUrl, ExtErpErrorResponse.class);
    }

    @Override
    public void close() {
        Closer.close(httpClient);
    }

    public Result<Void, ExtErpErrorResponse> checkHealth() {
        return httpClient.get(HEALTH_ENDPOINT);
    }

    public Result<ExtProductDetailsResponse, ExtErpErrorResponse> getProduct(String sku) {
        return httpClient.get(PRODUCTS_ENDPOINT + "/" + sku, ExtProductDetailsResponse.class);
    }
}
